src/csharp/Microsoft.Spark/Interop/Ipc/JsonSerDe.cs

```diff
@@ -2,70 +2,155 @@
+                    jsonElement.SortObjectProperties(writer);
+                    break;
+                case JsonValueKind.Array:
+                    jsonElement.SortArrayProperties(writer);
+                    break;
+                case JsonValueKind.String:
+                case JsonValueKind.Number:
+                case JsonValueKind.True:
+                case JsonValueKind.False:
+                case JsonValueKind.Null:
+                    jsonElement.WriteTo(writer);
+                    break;
+            };
+        }
+
+        /// <summary>
+        /// Sort properties that are Json objects.
+        /// </summary>
+        private static void SortObjectProperties(this JsonElement jObject, Utf8JsonWriter writer)
+        {
+            Debug.Assert(jObject.ValueKind == JsonValueKind.Object);
 
-            var sortedJArray = new JArray();
-            foreach (JToken item in jArray)
+            var propertyNames = new List<string>();
+            foreach (JsonProperty prop in jObject.EnumerateObject())
             {
-                if (item is JObject elem)
-                {
-                    sortedJArray.Add(elem.SortProperties());
-                }
-                else if (item is JArray arrayElem)
-                {
-                    sortedJArray.Add(arrayElem.SortProperties());
-                }
+                propertyNames.Add(prop.Name);
             }
+            propertyNames.Sort();
 
-            return sortedJArray;
+            writer.WriteStartObject();
+            foreach (string name in propertyNames)
+            {
+                writer.WritePropertyName(name);
+                jObject.GetProperty(name).WriteElementHelper(writer);
+            }
+            writer.WriteEndObject();
+        }  
+
+        /// <summary>
+        /// Extend method to sort items in a JSON array by keys.
+        /// </summary>
+        private static void SortArrayProperties(this JsonElement jArray, Utf8JsonWriter writer)
+        {
+            Debug.Assert(jArray.ValueKind == JsonValueKind.Array);
+
+            writer.WriteStartArray();
+            foreach (JsonElement item in jArray.EnumerateArray())
+            {
+                item.WriteElementHelper(writer);
+            }
+            writer.WriteEndArray();
         }
-    }
 
+        /// <summary>
+        /// Helper to appropriately write Json types based on their kind.
+        /// </summary>
+        private static void WriteElementHelper(this JsonElement item, Utf8JsonWriter writer)
+        {
+            Debug.Assert(item.ValueKind != JsonValueKind.Undefined);
+
+            if (item.ValueKind == JsonValueKind.Object)
+            {
+                item.SortObjectProperties(writer);
+            }
+            else if (item.ValueKind == JsonValueKind.Array)
+            {
+                item.SortArrayProperties(writer);
+            }
+            else
+            {
+                item.WriteTo(writer);
```

Comment:
Given what @imback82 said here (https://github.com/dotnet/spark/pull/358/files#r358410589), we should remove the else branch and not write anything here (for values within an array) in this PR.

That was the previous behavior. Maybe put a `Debug.Assert(item.ValueKind == JsonValueKind.Object || item.ValueKind == JsonValueKind.Array);`

If the scenario requires it, we should make this change in a separate PR.

Reply:
Removing that line breaks the functionality, as it's required for sorting objects. So, I've kept it and added a flag to toggle its use when sorting objects vs arrays. I've therefore left the Debug Assert as is since other types can come through.

Reply:
Good point. That works.